src/lib/statistics/join_graph_statistics_cache.hpp

```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include <map>
+#include <unordered_map>
+
+#include "boost/dynamic_bitset.hpp"
+
+#include "expression/abstract_expression.hpp"
+
+namespace opossum {
+
+class TableStatistics;
+class JoinGraph;
+
+/**
+ * Cache of TableStatistics for LQPs consisting exclusively of JoinNodes, PredicateNodes and
+ * <Vertex Nodes>. That is, this is a cache for statistics associated with LQPs representing a subset of a JoinGraph.
+ *
+ * The key of the cache is a bitmask (with each bit representing a predicate or a vertex node).
+ *
+ * This cache exists primarily to aid the performance of the JoinOrderingRule.
+ * The JOR frequently request statistics for different plans consisting of the same set of Join and Scan predicates.
+ */
+class JoinGraphStatisticsCache {
+ public:
+  using Bitmask = boost::dynamic_bitset<>;
+  using VertexIndexMap = std::unordered_map<std::shared_ptr<AbstractLQPNode>, size_t>;
+  using PredicateIndexMap = ExpressionUnorderedMap<size_t>;
+
+  static JoinGraphStatisticsCache from_join_graph(const JoinGraph& join_graph);
+
+  JoinGraphStatisticsCache(VertexIndexMap&& vertex_indices, PredicateIndexMap&& predicate_indices);
+
+  /**
+   * Try to build a bitmask (aka cache key) from an LQP. This will either return the bitmask or std::nullopt, if
+   * Predicates or Vertices not registered in _vertex_indices and _predicate_indices are encountered.
```

Comment:
Shouldn't be present, thinking about it. Put in asserts.

Reply:
Let's see whether they are triggered.

Reply:
Ah, I remember. Unexpected nodes can be encountered once the CardinalityEstimator "enters" a vertex subplan. 
No, I won't disable the cache in the CardinalityEstimator once that happens.

Reply:
Simply add that sentence?

Reply:
```c++
  /**
   * Try to build a bitmask (aka cache key) from an LQP. This will either return the bitmask or std::nullopt, if
   * Predicates or Vertices not registered in _vertex_indices and _predicate_indices are encountered. The latter can be
   * the case if `lqp` is a suplan of a vertex node.
   */
  std::optional<Bitmask> bitmask(const std::shared_ptr<AbstractLQPNode>& lqp) const;
```